### PR TITLE
ETB-96 SamV71 Serial CCSDS Driver skeleton

### PR DESCRIPTION
**Pull request by rphh**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+add_library(common_build_options INTERFACE)
+include(CommonBuildOptions)
+include(Sanitizers)
+
+option(TASTE_RUNTIME_COMMON_WARNINGS_AS_ERRORS
+       "Treat compiler wargnings as errors"
+       TRUE)
+
+if(OPTIONS_WARNINGS_AS_ERRORS)
```
Comment (Glamhoth) [inline CMakeLists.txt]: You've renamed the option, but left this check unchanged
Also building the code now throws some warning, so those have to be fixed

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                 -Wduplicated-branches
+                 -Wlogical-op)
+
+if(CMAKE_C_COMPILER_ID STREQUAL "Clang")
+    set(PROJECT_WARNINGS ${CLANG_WARNINGS})
+  elseif(CMAKE_C_COMPILER_ID STREQUAL "GNU")
+    set(PROJECT_WARNINGS ${GCC_WARNINGS})
+  endif()
+
+  target_compile_options(common_build_options INTERFACE ${PROJECT_WARNINGS})
```
Comment (Glamhoth) [inline CMakeLists.txt]: Unwanted indent

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake)
+
+include(MessageColors)
+include(LoggingHelper)
+
+add_library(common_build_options INTERFACE)
+include(CommonBuildOptions)
+include(Sanitizers)
+
+option(TASTE_RUNTIME_COMMON_WARNINGS_AS_ERRORS
```
Comment (Glamhoth) [inline CMakeLists.txt]: This should be `TASTE_SAMV71_DRIVERS_WARNINGS_AS_ERRORS`

File: src/app/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+                ${CMAKE_SOURCE_DIR}/TASTE-SAMV71-Runtime/src/BrokerLock/
+                ${CMAKE_SOURCE_DIR}/src/RuntimeMocks
+                ${CMAKE_SOURCE_DIR}/src
+                ${CMAKE_SOURCE_DIR}/TASTE-Runtime-Common/src/Escaper
+                ${CMAKE_SOURCE_DIR}/TASTE-Runtime-Common/src/Broker
+                ${CMAKE_SOURCE_DIR}/TASTE-Runtime-Common/src/Packetizer
+                ${CMAKE_SOURCE_DIR}/TASTE-Runtime-Common/src/RuntimeMocks
+                ${CMAKE_CURRENT_SOURCE_DIR})
+
+target_link_libraries(Samv71Demo
```
Comment (Glamhoth) [inline src/app/CMakeLists.txt:11]: missing `common_build_options`

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    assert("Not supported device name");
+  }
+}
+
+static inline void
+SamV71SerialCcsdsInit_uart_data_bits(samv71_serial_ccsds_private_data *self,
+                                     Serial_CCSDS_SamV71_Conf_T_bits bits) {
+  (void)self;
+  if (bits != 8) {
+    assert("Not supported number of data bits");
```
Comment (Glamhoth) [inline src/samv71_serial_ccsds/samv71_serial_ccsds.c]: assert(bits == 8 && "Not supported number of data bits")

File: src/app/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+target_include_directories(Samv71Demo
+    PRIVATE     ${CMAKE_SOURCE_DIR}/TASTE-SAMV71-Runtime/src
+                ${CMAKE_SOURCE_DIR}/TASTE-SAMV71-Runtime/src/BrokerLock/
+                ${CMAKE_SOURCE_DIR}/src/RuntimeMocks
+                ${CMAKE_SOURCE_DIR}/src
+                ${CMAKE_SOURCE_DIR}/TASTE-Runtime-Common/src/Escaper
+                ${CMAKE_SOURCE_DIR}/TASTE-Runtime-Common/src/Broker
+                ${CMAKE_SOURCE_DIR}/TASTE-Runtime-Common/src/Packetizer
+                ${CMAKE_SOURCE_DIR}/TASTE-Runtime-Common/src/RuntimeMocks
```
Comment (Glamhoth) [inline src/app/CMakeLists.txt]: Only `${CMAKE_SOURCE_DIR}/src` here is required

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+  Escaper_start_decoder(&self->m_escaper);
+  xSemaphoreTake(self->m_rx_semaphore, portMAX_DELAY);
+  Hal_uart_read(&self->m_hal_uart, self->m_fifo_memory_block,
+                Serial_CCSDS_SAMV71_RECV_BUFFER_SIZE, self->m_uart_rx_handler);
+
+  while (1) {
+    /// Wait for data to arrive. Semaphore will be given
+    xSemaphoreTake(self->m_rx_semaphore, portMAX_DELAY);
+    length = ByteFifo_getCount(&self->m_hal_uart.rxFifo);
+    if (length > 0) {
```
Comment (Glamhoth) [inline src/samv71_serial_ccsds/samv71_serial_ccsds.c]: Maybe `if (length <= 0) { return; }` for less indentation?

**Review by Glamhoth — CHANGES_REQUESTED** [review]

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "samv71_serial_ccsds.h"
+
+#include <assert.h>
+
+#include <EscaperInternal.h>
```
Comment (rphh) [inline src/samv71_serial_ccsds/samv71_serial_ccsds.c:27]: EscaperInternal.h in driver code? Is this necessary?

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+  SamV71SerialCcsdsInit_uart_register(self);
+  SamV71SerialCcsdsInit_uart_data_bits(self, device_configuration->bits);
+  SamV71SerialCcsdsInit_uart_parity(self, device_configuration->use_paritybit,
+                                    device_configuration->parity);
+  SamV71SerialCcsdsInit_uart_baudrate(self, device_configuration->speed);
+
+  Hal_uart_init(&self->m_hal_uart, self->m_hal_uart_config);
+}
+
+void UartRxCallback(void *private_data) {
```
Comment (rphh) [inline src/samv71_serial_ccsds/samv71_serial_ccsds.c]: static?

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+  samv71_serial_ccsds_private_data *self =
+      (samv71_serial_ccsds_private_data *)private_data;
+
+  xSemaphoreGiveFromISR(self->m_rx_semaphore, NULL);
+}
+
+ByteFifo *UartTxCallback(void *private_data) {
```
Comment (rphh) [inline src/samv71_serial_ccsds/samv71_serial_ccsds.c]: static?

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+  samv71_serial_ccsds_private_data *self =
+      (samv71_serial_ccsds_private_data *)private_data;
+  size_t length = 0;
+
+  Escaper_start_decoder(&self->m_escaper);
+  xSemaphoreTake(self->m_rx_semaphore, portMAX_DELAY);
+  Hal_uart_read(&self->m_hal_uart, self->m_fifo_memory_block,
+                Serial_CCSDS_SAMV71_RECV_BUFFER_SIZE, self->m_uart_rx_handler);
+
+  while (1) {
```
Comment (Lurkerpas) [inline src/samv71_serial_ccsds/samv71_serial_ccsds.c]: true

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    /// Wait for data to arrive. Semaphore will be given
+    xSemaphoreTake(self->m_rx_semaphore, portMAX_DELAY);
+    length = ByteFifo_getCount(&self->m_hal_uart.rxFifo);
+    if (length <= 0) {
+      return;
```
Comment (Lurkerpas) [inline src/samv71_serial_ccsds/samv71_serial_ccsds.c:202]: Log or/and assert

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c (hunk 8 above)
Comment (jrachucki-n7s) [reply src/samv71_serial_ccsds/samv71_serial_ccsds.c:27]: It is. It contains definition of STOP_BYTE used by the driver @rphh

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c (hunk 9 above)
Comment (jrachucki-n7s) [reply src/samv71_serial_ccsds/samv71_serial_ccsds.c]: fixed in new commit

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c (hunk 10 above)
Comment (jrachucki-n7s) [reply src/samv71_serial_ccsds/samv71_serial_ccsds.c]: fixed in new commit

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c (hunk 11 above)
Comment (jrachucki-n7s) [reply src/samv71_serial_ccsds/samv71_serial_ccsds.c]: fixed in new commit

File: src/samv71_serial_ccsds/samv71_serial_ccsds.c (hunk 12 above)
Comment (jrachucki-n7s) [reply src/samv71_serial_ccsds/samv71_serial_ccsds.c:202]: Added logging and assert in new commit

File: src/app/PacketSender.c [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+      (samv71_serial_ccsds_private_data *)args;
+  uint8_t packetData[PACKET_SIZE] = {0};
+
+  /* Initialise packet to be sent*/
+  CreatePacket(&packetizer, packetData);
+  /* Initialise xNextWakeTime - this only needs to be done once. */
+  xNextWakeTime = xTaskGetTickCount();
+  (void)xNextWakeTime;
+  for (;;) {
+    vTaskDelay(TASK_DELAY);
```
Comment (Lurkerpas) [inline src/app/PacketSender.c]: Does it have to use active polling?

Comment (jrachucki-n7s) [reply src/app/PacketSender.c]: No, it doesn't. But it was the quickest demonstration that I could implement. It can be changed if it is necessary.

**Review by Glamhoth — DISMISSED** [review]

File: src/app/PacketSender.h [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+#ifndef PACKET_SENDER_H
+#define PACKET_SENDER_H
+
+/**
+ * @file     PacketSender.h
+ * @brief    Part of samv71_serial_ccsds driver demo.
+ */
+
+/**
+ * @brief   Thread method. Initializes and sends packages
```
Comment (Lurkerpas) [inline src/app/PacketSender.h]: Packets

Comment (jrachucki-n7s) [reply src/app/PacketSender.h]: Fixed

**Review by Lurkerpas — APPROVED** [review]

**Pull request merged**